p2p/tests/node.rs

```diff
@@ -0,0 +1,220 @@
+                    recon_metrics.clone(),
+                ),
+                Recon::new(
+                    Arc::clone(&event_svc),
+                    ReconInterestProvider::new(node_key.id(), interest_svc),
+                    recon_metrics.clone(),
+                ),
+            )),
+            event_svc,
+            metrics,
+        )
+        .await?;
+        let cfg = iroh_rpc_client::Config {
+            p2p_addr: Some(rpc_client_addr),
+            channels: Some(1),
+            ..Default::default()
+        };
+
+        let client = iroh_rpc_client::Client::new(cfg).await?;
+
+        let network_events = p2p.network_events();
+        let task = tokio::task::spawn(async move { p2p.run().await.unwrap() });
+
+        let client = client.try_p2p()?;
+
+        let addr = tokio::time::timeout(Duration::from_millis(500), get_addr_loop(client.clone()))
+            .await
+            .context("timed out before getting a listening address for the node")??;
+        Ok(TestRunner {
+            task,
+            client,
+            peer_id,
+            network_events,
+            addr,
+        })
+    }
+}
+
+async fn get_addr_loop(client: P2pClient) -> Result<Multiaddr> {
+    loop {
+        let l = client.listeners().await?;
+        if let Some(a) = l.first() {
+            return Ok(a.clone());
+        }
+    }
+}
+
+struct TestRunner {
+    /// The task that runs the p2p node.
+    task: JoinHandle<()>,
+    /// The RPC client
+    /// Used to communicate with the p2p node.
+    client: P2pClient,
+    /// The node's peer_id
+    peer_id: PeerId,
+    /// A channel to read the network events received by the node.
+    network_events: Receiver<NetworkEvent>,
+    /// The listening address for this node.
+    addr: Multiaddr,
+}
+
+impl Drop for TestRunner {
+    fn drop(&mut self) {
+        self.task.abort();
+    }
+}
+
+#[tokio::test]
+async fn test_local_peer_id() -> Result<()> {
+    let test_runner = TestRunnerBuilder::new().build().await?;
+    let got_peer_id = test_runner.client.local_peer_id().await?;
+    assert_eq!(test_runner.peer_id, got_peer_id);
+    Ok(())
+}
+
+#[test(tokio::test(flavor = "multi_thread", worker_threads = 2))]
+async fn test_two_nodes() -> Result<()> {
+    let test_runner_a = TestRunnerBuilder::new().build().await?;
+    let test_runner_b = TestRunnerBuilder::new().build().await?;
+    let addrs_b = vec![test_runner_b.addr.clone()];
+    debug!(?test_runner_a.peer_id, ?test_runner_b.peer_id, "peer ids");
+
+    let peer_id_a = test_runner_a.client.local_peer_id().await?;
+    assert_eq!(test_runner_a.peer_id, peer_id_a);
+    let peer_id_b = test_runner_b.client.local_peer_id().await?;
+    assert_eq!(test_runner_b.peer_id, peer_id_b);
+
+    // connect
+    test_runner_a.client.connect(peer_id_b, addrs_b).await?;
+
+    // Make sure the peers have had time to negotiate protocols
+    tokio::time::sleep(Duration::from_millis(3_000)).await;
```

Comment:
It'd be nice if we could do this in a way that's event based somehow but if not this is fine.. maybe a short sleep and then a loop checking connected with a timeout? not a huge deal but it's nice to avoid any potential test flakiness if we can

Reply:
I had assumed it was not possible as the previous tests did this but I found it was quite simple. See newest commit.